File: src/main/java/com/github/ericjgagnon/vitest/run/VitestConsoleProperties.java
package com.github.ericjgagnon.vitest.run;

import com.intellij.execution.Executor;
import com.intellij.execution.process.ProcessHandler;
import com.intellij.execution.testframework.TestConsoleProperties;
import com.intellij.execution.testframework.actions.AbstractRerunFailedTestsAction;
import com.intellij.execution.testframework.sm.runner.OutputToGeneralTestEventsConverter;
import com.intellij.execution.testframework.sm.runner.SMTestLocator;
import com.intellij.execution.testframework.sm.runner.ui.SMTRunnerConsoleView;
import com.intellij.execution.ui.ConsoleView;
import com.intellij.javascript.nodejs.execution.NodeTargetRun;
import com.intellij.javascript.testing.JsTestConsoleProperties;
import com.intellij.terminal.TerminalExecutionConsole;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class VitestConsoleProperties extends JsTestConsoleProperties {
    public static final String TEST_FRAMEWORK_NAME = "VitestJavaScriptTestRunner";
    private final SMTestLocator locator;
    private final boolean withTerminalConsole;


    public VitestConsoleProperties(@NotNull VitestRunConfiguration configuration, @NotNull Executor executor, @NotNull SMTestLocator locator, boolean withTerminalConsole, @Nullable NodeTargetRun targetRun) {
        super(configuration, TEST_FRAMEWORK_NAME, executor, targetRun);
        this.locator = locator;
        this.withTerminalConsole = withTerminalConsole;
        this.setUsePredefinedMessageFilter(false);
        this.setIfUndefined(TestConsoleProperties.HIDE_PASSED_TESTS, false);
        this.setIfUndefined(TestConsoleProperties.HIDE_IGNORED_TEST, true);
        this.setIfUndefined(TestConsoleProperties.SCROLL_TO_SOURCE, true);
        this.setIfUndefined(TestConsoleProperties.SELECT_FIRST_DEFECT, true);
        this.setIdBasedTestTree(false);
        this.setPrintTestingStartedTime(false);
    }

    @Override
    public SMTestLocator getTestLocator() {
        return locator;
    }

    @NotNull
    public ConsoleView createConsole() {
        if (this.withTerminalConsole) {
            return new TerminalExecutionConsole(this.getProject(), null) {
                @Override
                public void attachToProcess(@NotNull ProcessHandler processHandler) {
                    super.attachToProcess(processHandler, false);
                }
            };
        } else {
            return super.createConsole();
        }
    }

    @Override
    public OutputToGeneralTestEventsConverter createTestEventsConverter(@NotNull String testFrameworkName, @NotNull TestConsoleProperties consoleProperties) {
        return super.createTestEventsConverter(testFrameworkName, consoleProperties);
    }

    @Override
    public @Nullable AbstractRerunFailedTestsAction createRerunFailedTestsAction(ConsoleView consoleView) {
        return new VitestRerunFailedTestAction((SMTRunnerConsoleView) consoleView, this);
    }
}

File: src/main/kotlin/com/github/ericjgagnon/vitest/run/VitestConsoleProperties.kt
package com.github.ericjgagnon.vitest.run

import com.github.ericjgagnon.vitest.run.VitestConstants.TEST_FRAMEWORK_NAME
import com.intellij.execution.Executor
import com.intellij.execution.process.ProcessHandler
import com.intellij.execution.testframework.actions.AbstractRerunFailedTestsAction
import com.intellij.execution.testframework.sm.runner.SMTestLocator
import com.intellij.execution.testframework.sm.runner.ui.SMTRunnerConsoleView
import com.intellij.execution.ui.ConsoleView
import com.intellij.javascript.nodejs.execution.NodeTargetRun
import com.intellij.javascript.testing.JsTestConsoleProperties
import com.intellij.terminal.TerminalExecutionConsole

class VitestConsoleProperties(
    configuration: VitestRunConfiguration,
    executor: Executor,
    private val locator: SMTestLocator,
    private val withTerminalConsole: Boolean,
    targetRun: NodeTargetRun?
) : JsTestConsoleProperties(configuration, TEST_FRAMEWORK_NAME, executor, targetRun) {
    init {
        this.isUsePredefinedMessageFilter = false
        setIfUndefined(HIDE_PASSED_TESTS, false)
        setIfUndefined(HIDE_IGNORED_TEST, true)
        setIfUndefined(SCROLL_TO_SOURCE, true)
        setIfUndefined(SELECT_FIRST_DEFECT, true)
        this.isIdBasedTestTree = false
        this.isPrintTestingStartedTime = false
    }

    override fun getTestLocator(): SMTestLocator? {
        return locator
    }

    override fun createConsole(): ConsoleView {
        return if (withTerminalConsole) {
            object : TerminalExecutionConsole(this.project, null) {
                override fun attachToProcess(processHandler: ProcessHandler) {
                    super.attachToProcess(processHandler, false)
                }
            }
        } else {
            super.createConsole()
        }
    }

    override fun createRerunFailedTestsAction(consoleView: ConsoleView): AbstractRerunFailedTestsAction? {
        return VitestRerunFailedTestAction((consoleView as SMTRunnerConsoleView), this)
    }
}